### PR TITLE
Enable prediction using Detectree2 model

### PR DESCRIPTION
**Pull request by amrithasp02**
This PR focuses on including Detectree2 model into the framework. All steps required to perform batch predictions using the dataloader have been included.

Main components include:
1. `Detectree2Detector` in `detector.py`
2. `Detectree2Module` in `models.py` - this class currently only deals with setting up the model config. When we implement the training portion, it will be used in a way similar to `DeepForestModule`.
3. `detectree2_detections.ipynb`

### CODE REVIEW

**Review by russelldj — COMMENTED** [review]
Looking good, just some cleanup suggestions.

After this is merged I'd be interested in thinking about how to simplify and reduce boilerplate as much as possible. It seems like it's a bit awkward having some of the functionality implemented in the `pl.Module` and other parts in the `Detector`. One thought I had was moving pretty much all of the functionality into the `Module` and only having one `LightningDetector` (not algorithm-specific) that generates predictions from that module. This would probably require each module implement an additional method, maybe called `geospatial_forward` or something like that, that takes in a batch from a torchgeo data loader and produces the predictions in the format that `predict_batch` currently returns. Then all of the pre- and post-processing that is currently done in `predict_batch` would be only in the `pl.Module`.

**Comment by russelldj** [comment]
Also, what do you think about adding this predictor as an option in the [predict detections](https://github.com/open-forest-observatory/tree-detection-framework/blob/main/examples/predict_detections.ipynb) example? You could provide the config constants for both in cell 2 as well as another one to select which model to use. And then in cell 5 create the appropriate detector.

File: tree_detection_framework/detection/detector.py [hunk 1]
```diff
@@ -415,3 +419,132 @@ def save_model(self, save_file: PATH_TYPE):
...
+                image = self.aug.get_transform(original_image).apply_image(
+                    original_image
+                )
+                image = torch.as_tensor(image.astype("float32").transpose(2, 0, 1))
+                image.to(self.cfg.MODEL.DEVICE)
+                print("Final image shape = ", image.shape)
+
+                input = {"image": image, "height": height, "width": width}
+                inputs.append(input)
+
```
[the remark this answers is not in the document]
Comment (russelldj) [reply tree_detection_framework/detection/detector.py:489]: This refers to the line below, sorry

File: tree_detection_framework/detection/models.py [hunk 2]
```diff
@@ -139,3 +143,92 @@ def configure_optimizers(self):
         # TODO: Return 'optimizer', 'lr_scheduler', 'monitor' when validation data is set
 
         return optimizer
+
+
+class Detectree2Module(lightning.LightningModule):
+    def __init__(self, param_dict: Optional[Dict[str, Any]] = None):
+        super().__init__()
+        # If param_dict is not provided, ensure it is an empty dictionary
+        self.param_dict = param_dict or {}
```
[the remark this answers is not in the document]
Comment (amrithasp02) [reply tree_detection_framework/detection/models.py:152]: Yes. I tried having {} in the function header, but I wasn't sure why it was throwing an error. That's why I had to add this.

File: tree_detection_framework/detection/detector.py [hunk 3]
```diff
@@ -415,3 +419,132 @@ def save_model(self, save_file: PATH_TYPE):
...
         # Should be implemented here
         raise NotImplementedError()
+
+
+class Detectree2Detector(LightningDetector):
+
+    def __init__(self, module):
+        self.module = module
+
+    def setup_model(self):
```
[the remark this answers is not in the document]
Comment (amrithasp02) [reply tree_detection_framework/detection/detector.py:429]: Since I have not yet implemented the trainer portion, I just kept it for now. But yeah, I can go ahead and remove the unimplemented functions. We can probably add it later.

File: tree_detection_framework/detection/detector.py [hunk 4]
```diff
@@ -415,3 +419,132 @@ def save_model(self, save_file: PATH_TYPE):
...
+
+    def setup_trainer(self):
+        raise NotImplementedError()
+
+    def setup_predictor(self):
+        """Build predictor model architecture and load model weights from config.
+        Based on `__init__` of `DefaultPredictor` from `detectron2`"""
+
+        self.cfg = self.module.cfg.clone()  # cfg can be modified by model
+        self.model = build_model(self.cfg)
```
[the remark this answers is not in the document]
Comment (amrithasp02) [reply tree_detection_framework/detection/detector.py:440]: `self.model` points to the detectron architecture whereas `self.module` points to the `LightningModule` which is currently unimplemented.

File: tree_detection_framework/detection/detector.py [hunk 5]
```diff
@@ -415,3 +419,132 @@ def save_model(self, save_file: PATH_TYPE):
...
+                original_image = original_image[:, :, :3]
+                print("Original image shape = ", original_image.shape)
+
+                height, width = original_image.shape[:2]
+                image = self.aug.get_transform(original_image).apply_image(
+                    original_image
+                )
+                image = torch.as_tensor(image.astype("float32").transpose(2, 0, 1))
+                image.to(self.cfg.MODEL.DEVICE)
+                print("Final image shape = ", image.shape)
```
[the remark this answers is not in the document]
Comment (amrithasp02) [reply tree_detection_framework/detection/detector.py:485]: Oops sorry, added it for debugging, and forgot to remove

File: tree_detection_framework/detection/models.py [hunk 6]
```diff
@@ -139,3 +143,92 @@ def configure_optimizers(self):
         # TODO: Return 'optimizer', 'lr_scheduler', 'monitor' when validation data is set
 
         return optimizer
+
+
+class Detectree2Module(lightning.LightningModule):
```
[the remark this answers is not in the document]
Comment (amrithasp02) [reply tree_detection_framework/detection/models.py:148]: This class isn't exactly implementing the `LightningModule` methods right now. Unlike `DeepForest` the existing code of `Detectron2` isn't using `pytorch lightning`. I wanted to know if we want to retain the original implementation or convert it to `pytorch lightning` before I proceeded to implement that.

[the remark this answers is not in the document]
Comment (russelldj) [reply tree_detection_framework/detection/models.py:148]: Sounds good to talk about this

**Comment by russelldj** [comment]
From the discussion, we need to make sure that this uses the GPU.

File: tree_detection_framework/detection/detector.py [hunk 7]
```diff
@@ -415,3 +419,115 @@ def save_model(self, save_file: PATH_TYPE):
...
+
+        Returns:
+            batch_preds (List[Dict[str, Instances]]): An iterable with a dictionary per image having "instances" value
+            as an `Instances` object containing prediction results.
+        """
+        self.setup_predictor()
+
+        with torch.no_grad():
+            inputs = []
+            for original_image in batch:
```
Comment (russelldj) [inline tree_detection_framework/detection/detector.py:472]: I missed this in the first review, is there any way to run inference on a batch of images at once? It would be nice to take advantage of the full size of the GPU if possible.

**Review by russelldj — APPROVED** [review]
Looking good. I just realized this doesn't do batched inference in a way that provides any acceleration. If it's not a simple fix to add this and you want wait until after you merge, can you create an issue to address it?

File: tree_detection_framework/detection/detector.py (hunk 7 above)
Comment (amrithasp02) [reply tree_detection_framework/detection/detector.py:472]: Thanks for catching that! I hadn't realized that passing the images this way could limit the GPU utilization. `Detectron2` is setup to handle images individually this way, passed as a list of dictionaries. I think this should be fixed when we bring in `pytorch lightning` and remove `Detectron2` as a "layer" inbetween. I'll create an issue for this.
